std/assembly/array.ts

```diff
@@ -476,6 +476,64 @@ export class Array<T> extends ArrayBufferView {
     return this;
   }
 
+  flat(): valueof<T>[] {
+    if (!isArray<T>()) {
+      ERROR("Cannot call flat() on Array<T> where T is not an Array.");
+    }
+    // Get the length and data start values
+    let length = this.length_;
+    let selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    let size: i32 = 0;
+    for (let i = 0; i < length; i++) {
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      size += child == null ? 1 : child.length;
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    let byteLength = size << usize(alignof<valueof<T>>());
+    let dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value
+    let result: valueof<T>[] = changetype<valueof<T>[]>(__alloc(offsetof<valueof<T>[]>(), idof<valueof<T>[]>()));
+    store<i32>(changetype<usize>(result), size, offsetof<valueof<T>[]>("length_"));
+    result.dataLength = byteLength;
+    result.dataStart = dataStart;
+    // this statement retains the ArrayBuffer
+    result.data = changetype<ArrayBuffer>(dataStart);
+
+    // set the elements
+    let resultIndex: i32 = -1;
+    for (let i = 0; i < length; i++) { // for each child
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      resultIndex++;
+
+      if (child == null) {
+        // if child == null and valueof<T> is a numeric type, it will push 0 here
+        store<valueof<T>>(dataStart + (<usize>resultIndex << usize(alignof<valueof<T>>())), null);
+        continue;
+      }
+
+      let childDataLength = child.dataLength;
+      let childDataStart = child.dataStart;
+      memory.copy(
+        dataStart + (<usize>resultIndex << usize(alignof<valueof<T>>())),
+        childDataStart,
+        childDataLength,
+      );
+
+      let childLength: i32 = (childDataLength >> usize(alignof<valueof<T>>()));
```

Comment:
Implicit conversion from usize to i32 will fail if usize is 64 bit.

```ts
let childLength = childDataLength >> u32(alignof<valueof<T>>());
```

Reply:
Okay.

Reply:
This caused the `j` parameter to become a `u32` as well. Fixed.